src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -463,6 +463,11 @@ private void Dispose(bool disposing)
                 {
                     Session.close(); // invoked Dispose()
                 }
+
+                if (Session != null && Session.graph != IntPtr.Zero)
+                {
+                    Session.graph.Dispose();
```

Comment:
Check out the comment at the top of this method, and this article: [https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-disposeboolean-overload](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-disposeboolean-overload)

I'm not sure we should be disposing other managed objects if this object is being finalized (i.e. disposing is `false`). That object might have already been finalized, which may lead to unintended errors.

Reply:
@eric.erhardt@microsoft.com We need to call dispose explicitly on the graph because it is not disposed by the destructor. TF .NET disposes the graph if you use their session layer which we don't. 

---
In reply to: [325323057](https://github.com/dotnet/machinelearning/pull/4223#discussion_r325323057) [](ancestors = 325323057)

Reply:
@eerhardt  We need to call dispose explicitly on the graph because it is not disposed by the destructor. TF .NET disposes the graph if you use their session layer which we don't. 
CC: @Oceania2018

Reply:
You shouldn't be touching other objects if `disposing` is `false`.

Please read 
https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-disposeboolean-overload 

and 

https://ericlippert.com/2015/05/21/when-everything-you-know-is-wrong-part-two/.  "Myth: An object being finalized can safely access another object."

> This myth follows directly from the previous. If you have a tree of objects and you are finalizing the root, then the children are still alive — because the root is alive, because it is on the finalization queue, and so the children have a living reference — but the children may have already been finalized, and are in no particularly good state to have their methods or data accessed.

___________

If the GC is finalizing this object, the `Session` may have gotten GC'd already, along with it's `.graph`. If that happens, and you try to dereference it here and call methods on it, this code will throw an exception and bring down the process. As stated in both those links, you can't access other objects in your finalizer (i.e. when `disposing` is `false`).

The `graph` object here should also implement a finalizer, and clean up unmanaged resources when it is finalized.

Reply:
I have read the article and also understand your point. Graph does not seem to define a destructor but that is ok, I can look into making this transform disposable so that when pipeline goes out of scope it calls dispose on this where we dispose the graph. @oceania2018 can we please have destructor in Graph that frees unmanaged memory?

Reply:
@codemzs `Graph` inherits from `DisposableObject`, so it has destructor. You just need to call `Graph.Dispose()` when you don't need it, that's it.